File: src/converter/plugins/convertTypeReference.ts
import ts from "typescript";
import {createSimplePlugin} from "../plugin.js";
import {convertNodeWithTypeArguments} from "./convertNodeWithTypeArguments.js";
import {CheckCoverageService, checkCoverageServiceKey} from "./CheckCoveragePlugin.js";

export const convertTypeReference = createSimplePlugin((node, context, render) => {
    if (!ts.isTypeReferenceNode(node)) return null

    const checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
    checkCoverageService?.cover(node)

    return render(node.typeName) + convertNodeWithTypeArguments.render(node, context, render)
})

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/convertTypeReference.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.createSimplePlugin
import typescript.Node
import typescript.isTypeReferenceNode

val convertTypeReference = createSimplePlugin plugin@{ node: Node, context, render ->
    if (!isTypeReferenceNode(node)) return@plugin null

    val checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
    checkCoverageService?.cover(node)

    render(node.typeName) + convertNodeWithTypeArguments.render(node, context, render)
}
